soc/arm/renesas_smartbond/da1469x/power.c

```diff
@@ -0,0 +1,153 @@
+#include <DA1469xAB.h>
+#include <da1469x_clock.h>
+#include <da1469x_otp.h>
+#include <da1469x_pd.h>
+#include <da1469x_pdc.h>
+#include <smartbond_sleep.h>
+
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+static int pdc_idx_combo;
+static int pdc_idx_sw_trigger;
+static bool wait_for_jtag;
+
+static bool z_smartbond_is_sleep_allowed(void)
+{
+	if (wait_for_jtag) {
+		if (CRG_TOP->SYS_STAT_REG & CRG_TOP_SYS_STAT_REG_DBG_IS_ACTIVE_Msk) {
+			wait_for_jtag = false;
+		}
+		return false;
+	}
+
+	/* We can enter extended sleep only if running from RCX or XTAL32K, debugger is
+	 * not attached and there are no interrupts pending.
+	 */
+	return (CRG_TOP->CLK_CTRL_REG & CRG_TOP_CLK_CTRL_REG_LP_CLK_SEL_Msk) &&
+	       !(CRG_TOP->SYS_STAT_REG & CRG_TOP_SYS_STAT_REG_DBG_IS_ACTIVE_Msk) &&
+	       !((NVIC->ISPR[0] & NVIC->ISER[0]) | (NVIC->ISPR[1] & NVIC->ISER[1]));
+}
+
+static bool z_smartbond_is_wakeup_by_jtag(void)
+{
+	return (da1469x_pdc_is_pending(pdc_idx_combo) &&
+		!(NVIC->ISPR[0] & ((1 << CMAC2SYS_IRQn) | (1 << KEY_WKUP_GPIO_IRQn) |
+				   (1 << VBUS_IRQn))));
+}
+
+static enum {
+	/** System not slept */
+	SLEEP_STATE_ACTIVE_STATE,
+	/** PD_SYS was released, deep sleep allowed */
+	SLEEP_STATE_PD_SYS_RELEASED,
+	/** System returned from deep sleep */
+	SLEEP_STATE_SLEPT,
+} sleep_state;
+
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(substate_id);
+
+	switch (state) {
+	case PM_STATE_STANDBY:
+		/* We enter here with interrupts disabled by BASEPRI which prevents wfi from
+		 * waking up on interrupts. Need to disable interrupts by PRIMASK instead and
+		 * reset BASEPRI to 0.
+		 */
+		__disable_irq();
+		arch_irq_unlock(0);
+
+		if (!z_smartbond_is_sleep_allowed()) {
+			barrier_dsync_fence_full();
+			__WFI();
+			break;
+		}
+
+		da1469x_pdc_set(pdc_idx_sw_trigger);
+
+		sleep_state = SLEEP_STATE_PD_SYS_RELEASED;
+		/* PD_SYS will not be disabled here until we enter deep sleep - don't wait */
+		if (!da1469x_pd_release_nowait(MCU_PD_DOMAIN_SYS)) {
+			barrier_dsync_fence_full();
+			__WFI();
+		} else {
+			da1469x_pdc_ack_all_m33();
+			if (z_smartbond_sleep()) {
+				sleep_state = SLEEP_STATE_SLEPT;
+			}
+		}
+
+		break;
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
+}
+
+__weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(state);
+	ARG_UNUSED(substate_id);
+
+	if (sleep_state == SLEEP_STATE_SLEPT) {
+		if (!IS_ENABLED(CONFIG_WDT_SMARTBOND)) {
+			/* Watchdog is always resumed when PD_SYS is turned off, need to
+			 * freeze it again if there's no one to feed it.
+			 */
+			GPREG->SET_FREEZE_REG = GPREG_SET_FREEZE_REG_FRZ_SYS_WDOG_Msk;
+			SYS_WDOG->WATCHDOG_REG = SYS_WDOG_WATCHDOG_REG_WDOG_VAL_Msk;
+		}
+		sleep_state = SLEEP_STATE_ACTIVE_STATE;
+
+		da1469x_pd_acquire(MCU_PD_DOMAIN_SYS);
+
+		if (z_smartbond_is_wakeup_by_jtag()) {
+			wait_for_jtag = 1;
+		}
+
+		if (DT_SAME_NODE(DT_PROP(DT_NODELABEL(sys_clk), clock_src),
```

Comment:
@kasjer I want to clarify this - system PM will first call `pm_resume_devices` and then `pm_state_exit_post_ops` where the system clock is switched from `RC32M` --> `XTAL32M` (after exiting the sleep state). It might happen that a device driver requests PLL from within its registered PM callback. In such a case, an assertion should be thrown since PLL is requested while XTAL32M has yet to be enabled (which is not allowed). To my view there are three approaches to overcome this:
 - Enabling XTAL32M is moved to `pm_state_set` right after `sleep_state = SLEEP_STATE_SLEPT` (before calling `pm_resume_devices`)
 - A device driver should first check that XTAL32M is up and running before switching to PLL (this should be documented) 
 - Following to enabling XTAL32M, we also track whether the device was running @ PLL so the last selected clock is re-applied. In doing so, a device driver should not have to check the current clock that might have been changed due to the device entering the sleep state.

To my view, the last option is the best approach.

Reply:
@ioannis-karachalios we could fix PLL handling later. I only tested without PLL, since.

Reply:
OK. We should revisit this part at a next round.